### PR TITLE
New tcp proxy

### PR DESCRIPTION
**Pull request by dberliner**

### CODE REVIEW

File: drivers/usb/gadget/xaptum/f_psock_gadget.c [hunk 1]
```diff
@@ -89,6 +100,76 @@ void f_psock_gadget_work_handler( struct work_struct *work )
  * USB DESCRIPTOR DEFINITIONS
  ***************************************************************************/
 
+/* Control Endpoint Descriptions. 
+ * These descriptors do not vary by speed setting becuase FS, HS and SS all 
+ * have lengths >=64 bytes, and SCM protocol standards limit control 
+ * transfer packets to 64 bytes.
+ */
+static struct usb_endpoint_descriptor
```
Comment (dberliner) [inline drivers/usb/gadget/xaptum/f_psock_gadget.c:143]: Interrupt endpoint descriptors. Compare to https://elixir.bootlin.com/linux/latest/source/drivers/usb/gadget/function/f_ecm.c#L192

File: drivers/usb/gadget/xaptum/f_psock_gadget.c [hunk 2]
```diff
@@ -224,23 +327,87 @@ static struct usb_gadget_strings *psock_strings[] = {
...
  */
 
 static inline struct f_psock *func_to_psock(struct usb_function *f)
 {
+	printk(KERN_INFO "func_to_psock enter");
         return container_of(f, struct f_psock, function);
 }
 
 
+/* Binds this driver to a device */
```
Comment (dberliner) [inline drivers/usb/gadget/xaptum/f_psock_gadget.c:423]: This functions sets up the endpoints and successfully runs before the crash. The following is outputted

```
psock_bind enter
Configured cmd_out_ep
Configured cmd_in_ep
psock_bind exit ret=0
```

File: drivers/usb/gadget/xaptum/f_psock_gadget.c [hunk 3]
```diff
@@ -89,21 +100,91 @@ void f_psock_gadget_work_handler( struct work_struct *work )
...
+};
+
 /*
  * usb interface descriptor
  */
 static struct usb_interface_descriptor psock_intf = {
 	.bLength = sizeof(psock_intf),
 	.bDescriptorType = USB_DT_INTERFACE,
-	.bNumEndpoints = 2,
+	.bNumEndpoints = 4,
```
Comment (dberliner) [inline drivers/usb/gadget/xaptum/f_psock_gadget.c]: This was the cause of a crash. Using the device on hardware circumvented the crash and the host (Advantech gateway) was able to report a mismatch in the descriptor count that lead me to this...

Comment (drbild) [reply drivers/usb/gadget/xaptum/f_psock_gadget.c]: nice!

**Review by drbild — CHANGES_REQUESTED** [review]
As with the host driver, could you pare this back to just the code required to create the device and endpoints.

Leave everything for sending and receiving actual messages to a later PR.

**Comment by dberliner** [comment]
I want to keep this branch so I'll open a new PR.

**Pull request closed without merging**